### PR TITLE
feat(cli): `cdk rollback`

### PR DESCRIPTION
**Pull request by rix0rrr**
Add a CLI feature to roll a stuck change back.

This is mostly useful for deployments performed using `--no-rollback`: if a failure occurs, the stack gets stuck in an `UPDATE_FAILED` state from which there are 2 options:

- Try again using a new template
- Roll back to the last stable state

There used to be no way to perform the second operation using the CDK CLI, but there now is.

`cdk rollback` works in 2 situations:

- A paused fail state; it will initiating a fresh rollback (on `CREATE_FAILED`, `UPDATE_FAILED`).
- A paused rollback state; it will retry the rollback, optionally skipping some resources (on `UPDATE_ROLLBACK_FAILED` -- it seems there is no way to continue a rollback in `ROLLBACK_FAILED` state).

`cdk rollback --orphan <logicalid>` can be used to skip resource rollbacks that are causing problems.

`cdk rollback --force` will look up all failed resources and continue skipping them until the rollback has finished.

This change requires new bootstrap permissions, so the bootstrap stack is updated to add the following IAM permissions to the `deploy-action` role:

```
                  - cloudformation:RollbackStack
                  - cloudformation:ContinueUpdateRollback
```

These are necessary to call the 2 CloudFormation APIs that start and continue a rollback. 

Relates to (but does not close yet) #30546.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by rix0rrr** [comment]
This PR is currently blocked by `cdk-assets` being on an old version of `@aws-cdk/cloud-assembly-schema`, while the framework is on a new version. Hence, integ tests will fail.

File: packages/aws-cdk/lib/api/util/cloudformation/stack-status.ts
Comment (iliapolo) [inline packages/aws-cdk/lib/api/util/cloudformation/stack-status.ts:64]: I think this should cause an error because it doesn't achieve what the user wanted.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (iliapolo) [inline packages/aws-cdk/lib/api/deployments.ts:523]: The message is misleading, it might be that the stack cannot be rolled back, in which case I think we should error out because we didn't get the user to their desired state.

File: packages/aws-cdk/README.md
Comment (rix0rrr) [reply packages/aws-cdk/README.md:210]: I don't know. My guess is that the storage that holds `(old resource, new resource)` can't outlast a single workflow execution. I guess there's no reason it couldn't be part of the long-term stack state, so that a rollback could delete the new resource and stabilize on the old one. But that's not how it's done.

File: packages/aws-cdk/lib/api/cxapp/exec.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/cxapp/exec.ts:52]: I ended up with this during debugging. Everything is very loosely typed, and it ended up as an array for me. I forgot why that was exactly.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/deployments.ts:22]: Yep

File: packages/aws-cdk/lib/api/deployments.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/deployments.ts:508]: Not sure what you mean. How does it downgrade what?

File: packages/aws-cdk/lib/api/deployments.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/deployments.ts:523]: This was for the case where the app has multiple stacks. We shouldn't require that ALL of them are rollbackable, we just want to get back to a stable state.

But this reminded me that I had the plan to add something to the driving loop that I forgot, so I'll add the condition there that at least one stack needs to be in rollbackable state.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/deployments.ts:282]: It's used now

File: packages/aws-cdk/lib/api/deployments.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/deployments.ts:603]: Since this is an interactive command, I'm not too concerned.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/deployments.ts:531]: We might not, but we should.

The token is there so that when the SDK retries because of network trouble, a command that has successfully been received by the server (but its success response hasn't made it to the client yet) will not be performed again (and in this particular case produce an error response, even though the rollback was started successfully)

File: packages/aws-cdk/lib/cli.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/cli.ts:182]: Makes sense. "Force" is the term that CloudFormation itself uses for this function.

File: packages/aws-cdk/lib/cdk-toolkit.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/cdk-toolkit.ts:466]: What do you mean?

File: packages/aws-cdk/test/util/stack-monitor.test.ts
Comment (rix0rrr) [reply packages/aws-cdk/test/util/stack-monitor.test.ts:135]: Yeah. The code that polls for CloudFormation events and determines whether certain events are errors, etc, was locked up in the activity monitor. It would only periodically poll and then print the results.

I also wanted it on-demand for the `--force` (`--orphan-all`) feature, so I pulled it out into a separate class.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/deployments.ts:594]: The code there is really quite geared towards deployments and nothing else. There isn't that much functionality in that code that it's worth refactoring (it mostly just stitches together other code).

I agree that a bigger refactor of that whole area of code might be worthwhile, but I don't know that this PR is the best one to do it in.

File: packages/aws-cdk/README.md
Comment (iliapolo) [reply packages/aws-cdk/README.md:210]: I tried it and it worked...I might have missed something though. Anyway, we can punt this.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (iliapolo) [reply packages/aws-cdk/lib/api/deployments.ts:514]: If a future stack defines `requiresBootstrapStackVersion: 25` - this would still check for `23`, and run the rollback (i.e stack deployment), bypassing the check.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (iliapolo) [reply packages/aws-cdk/lib/api/deployments.ts:604]: Even if it is interactive, expecting the user to hit `Ctrl+C` to terminate is not a great experience.

File: packages/aws-cdk/lib/cdk-toolkit.ts
Comment (iliapolo) [reply packages/aws-cdk/lib/cdk-toolkit.ts:469]: Oh so I think I was commenting from the position that if a stack cannot be rolled back, we should throw an error - in this case a single faulty stack can prevent rolling back others, so we need to accumulate errors. 

If you insist on no-oping for a unrollable stack, this is fine, but I still think we should error out.

File: packages/aws-cdk/lib/cli.ts
Comment (iliapolo) [reply packages/aws-cdk/lib/cli.ts:182]: Got it. Your call then.

File: packages/aws-cdk/README.md
Comment (rix0rrr) [reply packages/aws-cdk/README.md:210]: It would surprise me, since all the CFN docs say it doesn't work. Doesn't mean the docs aren't out of date though.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/deployments.ts:529]: I see. You would prefer `throw new NotRollbackeableError()` and then catch specifically that, so that this function can happily fail

File: packages/@aws-cdk-testing/cli-integ/resources/cdk-apps/rollback-test-app/app.js
Comment (iliapolo) [inline packages/@aws-cdk-testing/cli-integ/resources/cdk-apps/rollback-test-app/app.js:11]: Needs to be adjusted for phases `2a` and `2b`. I'm good with just removing this and let the tests speak for themselves.

File: packages/aws-cdk/lib/api/deployments.ts
Comment (iliapolo) [reply packages/aws-cdk/lib/api/deployments.ts:514]: Fair enough.

File: packages/aws-cdk/lib/cdk-toolkit.ts
Comment (iliapolo) [reply packages/aws-cdk/lib/cdk-toolkit.ts:469]: So I see that right now you do throw an error on `ROLLBACK_FAILED` state - doesn't this mean you need to swallow the error here and proceed? to avoid the poison pill?

**Review by iliapolo — APPROVED** [review]

**Pull request merged**
